### PR TITLE
✨ feat(rust/tree): add binary_tree_bfs, binary_tree_dfs, 

### PR DESCRIPTION
**Pull request by xBLACKICEx**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

image
image

### CODE REVIEW

**Comment by krahets** [comment]
To be reviewed by @sjinzh

**Review by sjinzh — APPROVED** [review]
@krahets It's OK.

File: codes/rust/include/tree_node.rs [hunk 1]
```diff
@@ -26,4 +27,65 @@ impl TreeNode {
...
+    let mut que = VecDeque::new();
+    que.push_back(Rc::clone(&root));
+
+    let mut index = 0;
+    while let Some(node) = que.pop_front() {
+        index += 1;
+        if index >= list.len() {
+            break;
+        }
+        if index < list.len() {
```
Comment (krahets) [inline codes/rust/include/tree_node.rs:76]: Incorrect. Please refer to the other language implementation

File: codes/rust/include/tree_node.rs [hunk 2]
```diff
@@ -26,4 +27,65 @@ impl TreeNode {
...
+            node.borrow_mut().left = Some(TreeNode::new(list[index]));
+            que.push_back(Rc::clone(&node.borrow().left.as_ref().unwrap()));
+        }
+
+        index += 1;
+        if index >= list.len() {
```
Comment (krahets) [inline codes/rust/include/tree_node.rs:82]: Same as above

File: codes/rust/include/tree_node.rs (hunk 1 above)
Comment (xBLACKICEx) [reply codes/rust/include/tree_node.rs:76]: Sorry, I can't find out what you said is Incorrect，from other languages. are you mean changing list.len () to variable n?

Comment (krahets) [reply codes/rust/include/tree_node.rs:76]: https://github.com/krahets/hello-algo/blob/2351fd1a1a0d73410ac3537df79d293bf2e205e9/codes/python/modules/binary_tree.py#L33

The function is designed to convert the list to any trees. So we should parse the null node from the list.
This line determines whether the node is null so we can handle the list like `[1, None, 2, 3, ...]`.

**Comment by xBLACKICEx** [comment]
It should be ok by now,

**Comment by xBLACKICEx** [comment]
It should be ok by now, i try some test output is same as python.

**Review by krahets — APPROVED** [review]
Thank @xBLACKICEx for the PR!

**Pull request merged**